File: integration_tests/java_client_tests/client-app/src/test/java/dev/graphlink/test/InputsTest.java
package dev.graphlink.test;

import dev.graphlink.test.generated.client.GraphLinkClient;
import dev.graphlink.test.generated.enums.Priority;
import dev.graphlink.test.generated.enums.UserStatus;
import dev.graphlink.test.generated.inputs.AddressInput;
import dev.graphlink.test.generated.inputs.CreateUserInput;
import dev.graphlink.test.generated.inputs.UpdateUserInput;
import dev.graphlink.test.generated.types.CreateUserResponse;
import dev.graphlink.test.generated.types.UpdateUserResponse;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import java.util.List;
import java.util.Map;

import static dev.graphlink.test.Fixtures.*;
import static org.junit.jupiter.api.Assertions.*;

class InputsTest {

    private MockAdapter adapter;
    private GraphLinkClient client;

    private static final AddressInput kAddressInput = AddressInput.builder()
            .street("123 Main St")
            .city("Springfield")
            .country("US")
            .zip("12345")
            .build();

    @BeforeEach
    void setUp() {
        adapter = new MockAdapter();
        client = newClient(adapter, new MockWebSocketAdapter());
    }

    // ── CreateUserInput — required fields ─────────────────────────────────────

    @Nested
    class CreateUserInput_RequiredFields {

        private CreateUserInput minimalInput;

        @BeforeEach
        void setup() {
            adapter.registerData("createUser", Map.of("createUser", kUserAliceJson));
            minimalInput = CreateUserInput.builder()
                    .name("Alice Smith")
                    .email("alice@test.com")
                    .status(UserStatus.ACTIVE)
                    .address(kAddressInput)
                    .build();
        }

        @Test
        void createUser_sendsNameInVariables() {
            client.mutations.createUser(minimalInput);
            @SuppressWarnings("unchecked")
            Map<String, Object> inputVar = (Map<String, Object>) adapter.getLastCall().variables().get("input");
            assertEquals("Alice Smith", inputVar.get("name"));
        }

        @Test
        void createUser_sendsStatusAsString() {
            client.mutations.createUser(minimalInput);
            @SuppressWarnings("unchecked")
            Map<String, Object> inputVar = (Map<String, Object>) adapter.getLastCall().variables().get("input");
            assertEquals("ACTIVE", inputVar.get("status"));
        }

        @Test
        void createUser_sendsAddressNested() {
            client.mutations.createUser(minimalInput);
            @SuppressWarnings("unchecked")
            Map<String, Object> inputVar = (Map<String, Object>) adapter.getLastCall().variables().get("input");
            @SuppressWarnings("unchecked")
            Map<String, Object> addrVar = (Map<String, Object>) inputVar.get("address");
            assertEquals("Springfield", addrVar.get("city"));
        }

        @Test
        void createUser_nullableFields_areNull_whenNotSet() {
            client.mutations.createUser(minimalInput);
            @SuppressWarnings("unchecked")
            Map<String, Object> inputVar = (Map<String, Object>) adapter.getLastCall().variables().get("input");
            assertNull(inputVar.get("priority"));
            assertNull(inputVar.get("billingAddress"));
            assertNull(inputVar.get("tags"));
        }

        @Test
        void createUser_response_deserializesCorrectly() {
            CreateUserResponse res = client.mutations.createUser(minimalInput);
            assertNotNull(res.getCreateUser());
            assertEquals("user-1", res.getCreateUser().getId());
        }
    }

    // ── CreateUserInput — optional fields ─────────────────────────────────────

    @Nested
    class CreateUserInput_OptionalFields {

        @BeforeEach
        void register() {
            adapter.registerData("createUser", Map.of("createUser", kUserAliceJson));
        }

        @Test
        void createUser_sendsPriority_whenSet() {
            CreateUserInput input = CreateUserInput.builder()
                    .name("Alice Smith")
                    .email("alice@test.com")
                    .status(UserStatus.ACTIVE)
                    .address(kAddressInput)
                    .priority(Priority.HIGH)
                    .build();
            client.mutations.createUser(input);
            @SuppressWarnings("unchecked")
            Map<String, Object> inputVar = (Map<String, Object>) adapter.getLastCall().variables().get("input");
            assertEquals("HIGH", inputVar.get("priority"));
        }

        @Test
        void createUser_sendsTags_whenSet() {
            CreateUserInput input = CreateUserInput.builder()
                    .name("Alice Smith")
                    .email("alice@test.com")
                    .status(UserStatus.ACTIVE)
                    .address(kAddressInput)
                    .tags(List.of("admin", "beta"))
                    .build();
            client.mutations.createUser(input);
            @SuppressWarnings("unchecked")
            Map<String, Object> inputVar = (Map<String, Object>) adapter.getLastCall().variables().get("input");
            assertEquals(List.of("admin", "beta"), inputVar.get("tags"));
        }

        @Test
        void createUser_sendsInitialScore_whenSet() {
            CreateUserInput input = CreateUserInput.builder()
                    .name("Alice Smith")
                    .email("alice@test.com")
                    .status(UserStatus.ACTIVE)
                    .address(kAddressInput)
                    .initialScore(100)
                    .build();
            client.mutations.createUser(input);
            @SuppressWarnings("unchecked")
            Map<String, Object> inputVar = (Map<String, Object>) adapter.getLastCall().variables().get("input");
            assertEquals(100, inputVar.get("initialScore"));
        }
    }

    // ── UpdateUserInput — all nullable fields ─────────────────────────────────

    @Nested
    class UpdateUserInput_AllNullable {

        @BeforeEach
        void register() {
            adapter.registerData("updateUser", Map.of("updateUser", kUserAliceJson));
        }

        @Test
        void updateUser_emptyInput_sendsNullFields() {
            UpdateUserInput input = UpdateUserInput.builder().build();
            client.mutations.updateUser("user-1", input);
            @SuppressWarnings("unchecked")
            Map<String, Object> inputVar = (Map<String, Object>) adapter.getLastCall().variables().get("input");
            assertNull(inputVar.get("name"));
            assertNull(inputVar.get("status"));
            assertNull(inputVar.get("priority"));
        }

        @Test
        void updateUser_sendsIdInVariables() {
            UpdateUserInput input = UpdateUserInput.builder().build();
            client.mutations.updateUser("user-1", input);
            assertEquals("user-1", adapter.getLastCall().variables().get("id"));
        }

        @Test
        void updateUser_withStatus_sendsStatusAsString() {
            UpdateUserInput input = UpdateUserInput.builder()
                    .status(UserStatus.SUSPENDED)
                    .build();
            client.mutations.updateUser("user-1", input);
            @SuppressWarnings("unchecked")
            Map<String, Object> inputVar = (Map<String, Object>) adapter.getLastCall().variables().get("input");
            assertEquals("SUSPENDED", inputVar.get("status"));
        }

        @Test
        void updateUser_response_deserializesCorrectly() {
            UpdateUserInput input = UpdateUserInput.builder().name("Alice Updated").build();
            UpdateUserResponse res = client.mutations.updateUser("user-1", input);
            assertNotNull(res.getUpdateUser());
        }
    }

    // ── AddressInput — nullable zip ───────────────────────────────────────────

    @Nested
    class AddressInput_NullableZip {

        @Test
        void addressInput_zip_isNull_whenNotSet() {
            AddressInput addr = AddressInput.builder()
                    .street("1 Road")
                    .city("Town")
                    .country("US")
                    .build();
            assertNull(addr.getZip());
        }

        @Test
        void addressInput_zip_isPresent_whenSet() {
            AddressInput addr = AddressInput.builder()
                    .street("1 Road")
                    .city("Town")
                    .country("US")
                    .zip("00001")
                    .build();
            assertEquals("00001", addr.getZip());
        }
    }
}

File: integration_tests/kotlin_client_tests/client-app/src/test/kotlin/dev/graphlink/test/InputsTest.kt
package dev.graphlink.test

import dev.graphlink.test.generated.client.GraphLinkClient
import dev.graphlink.test.generated.enums.Priority
import dev.graphlink.test.generated.enums.UserStatus
import dev.graphlink.test.generated.inputs.AddressInput
import dev.graphlink.test.generated.inputs.CreateUserInput
import dev.graphlink.test.generated.inputs.UpdateUserInput
import kotlinx.coroutines.test.runTest
import org.junit.jupiter.api.Nested
import kotlin.test.*
import dev.graphlink.test.Fixtures.kUserAliceJson
import dev.graphlink.test.Fixtures.newClient

class InputsTest {

    private lateinit var adapter: MockAdapter
    private lateinit var client: GraphLinkClient

    private val kAddressInput = AddressInput("123 Main St", "Springfield", "US", "12345")

    @BeforeTest
    fun setUp() {
        adapter = MockAdapter()
        client = newClient(adapter, MockWebSocketAdapter())
    }

    // ── CreateUserInput — required fields ─────────────────────────────────────

    @Nested
    inner class CreateUserInput_RequiredFields {

        private val minimalInput = CreateUserInput(
            name = "Alice Smith",
            email = "alice@test.com",
            status = UserStatus.ACTIVE,
            address = AddressInput("123 Main St", "Springfield", "US", "12345"),
        )

        @BeforeTest
        fun register() {
            adapter.registerData("createUser", mapOf("createUser" to kUserAliceJson))
        }

        @Test
        fun createUser_sendsNameInVariables() = runTest {
            client.mutations.createUser(minimalInput)
            @Suppress("UNCHECKED_CAST")
            val inputVar = adapter.getLastCall()!!.variables["input"] as Map<String, Any?>
            assertEquals("Alice Smith", inputVar["name"])
        }

        @Test
        fun createUser_sendsStatusAsString() = runTest {
            client.mutations.createUser(minimalInput)
            @Suppress("UNCHECKED_CAST")
            val inputVar = adapter.getLastCall()!!.variables["input"] as Map<String, Any?>
            assertEquals("ACTIVE", inputVar["status"])
        }

        @Test
        fun createUser_sendsAddressNested() = runTest {
            client.mutations.createUser(minimalInput)
            @Suppress("UNCHECKED_CAST")
            val inputVar = adapter.getLastCall()!!.variables["input"] as Map<String, Any?>
            @Suppress("UNCHECKED_CAST")
            val addrVar = inputVar["address"] as Map<String, Any?>
            assertEquals("Springfield", addrVar["city"])
        }

        @Test
        fun createUser_nullableFields_areNull_whenNotSet() = runTest {
            client.mutations.createUser(minimalInput)
            @Suppress("UNCHECKED_CAST")
            val inputVar = adapter.getLastCall()!!.variables["input"] as Map<String, Any?>
            assertNull(inputVar["priority"])
            assertNull(inputVar["billingAddress"])
            assertNull(inputVar["tags"])
        }

        @Test
        fun createUser_response_deserializesCorrectly() = runTest {
            val res = client.mutations.createUser(minimalInput)
            assertNotNull(res.createUser)
            assertEquals("user-1", res.createUser.id)
        }
    }

    // ── CreateUserInput — optional fields ─────────────────────────────────────

    @Nested
    inner class CreateUserInput_OptionalFields {

        @BeforeTest
        fun register() {
            adapter.registerData("createUser", mapOf("createUser" to kUserAliceJson))
        }

        @Test
        fun createUser_sendsPriority_whenSet() = runTest {
            val input = CreateUserInput(
                name = "Alice Smith", email = "alice@test.com",
                status = UserStatus.ACTIVE, address = kAddressInput,
                priority = Priority.HIGH,
            )
            client.mutations.createUser(input)
            @Suppress("UNCHECKED_CAST")
            val inputVar = adapter.getLastCall()!!.variables["input"] as Map<String, Any?>
            assertEquals("HIGH", inputVar["priority"])
        }

        @Test
        fun createUser_sendsTags_whenSet() = runTest {
            val input = CreateUserInput(
                name = "Alice Smith", email = "alice@test.com",
                status = UserStatus.ACTIVE, address = kAddressInput,
                tags = listOf("admin", "beta"),
            )
            client.mutations.createUser(input)
            @Suppress("UNCHECKED_CAST")
            val inputVar = adapter.getLastCall()!!.variables["input"] as Map<String, Any?>
            assertEquals(listOf("admin", "beta"), inputVar["tags"])
        }

        @Test
        fun createUser_sendsInitialScore_whenSet() = runTest {
            val input = CreateUserInput(
                name = "Alice Smith", email = "alice@test.com",
                status = UserStatus.ACTIVE, address = kAddressInput,
                initialScore = 100,
            )
            client.mutations.createUser(input)
            @Suppress("UNCHECKED_CAST")
            val inputVar = adapter.getLastCall()!!.variables["input"] as Map<String, Any?>
            assertEquals(100, (inputVar["initialScore"] as Number).toInt())
        }
    }

    // ── UpdateUserInput — all nullable fields ─────────────────────────────────

    @Nested
    inner class UpdateUserInput_AllNullable {

        @BeforeTest
        fun register() {
            adapter.registerData("updateUser", mapOf("updateUser" to kUserAliceJson))
        }

        @Test
        fun updateUser_emptyInput_sendsNullFields() = runTest {
            client.mutations.updateUser("user-1", UpdateUserInput())
            @Suppress("UNCHECKED_CAST")
            val inputVar = adapter.getLastCall()!!.variables["input"] as Map<String, Any?>
            assertNull(inputVar["name"])
            assertNull(inputVar["status"])
            assertNull(inputVar["priority"])
        }

        @Test
        fun updateUser_sendsIdInVariables() = runTest {
            client.mutations.updateUser("user-1", UpdateUserInput())
            assertEquals("user-1", adapter.getLastCall()!!.variables["id"])
        }

        @Test
        fun updateUser_withStatus_sendsStatusAsString() = runTest {
            client.mutations.updateUser("user-1", UpdateUserInput(status = UserStatus.SUSPENDED))
            @Suppress("UNCHECKED_CAST")
            val inputVar = adapter.getLastCall()!!.variables["input"] as Map<String, Any?>
            assertEquals("SUSPENDED", inputVar["status"])
        }

        @Test
        fun updateUser_response_deserializesCorrectly() = runTest {
            val res = client.mutations.updateUser("user-1", UpdateUserInput(name = "Alice Updated"))
            assertNotNull(res.updateUser)
        }
    }

    // ── AddressInput — nullable zip ───────────────────────────────────────────

    @Nested
    inner class AddressInput_NullableZip {

        @Test
        fun addressInput_zip_isNull_whenNotSet() {
            val addr = AddressInput(street = "1 Road", city = "Town", country = "US")
            assertNull(addr.zip)
        }

        @Test
        fun addressInput_zip_isPresent_whenSet() {
            val addr = AddressInput(street = "1 Road", city = "Town", country = "US", zip = "00001")
            assertEquals("00001", addr.zip)
        }
    }
}
